File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        unordered_map<int,vector<int>> m;
        queue<pair<int,int>> q;
        q.push({headID,0});
        for(int i=0;i<manager.size();i++){
            m[manager[i]].push_back(i);
        }
        int maxTimeTaken = 0;
        while(!q.empty()){
            int n = q.size();
            for(int i=0;i<n;i++){
                int head = q.front().first;
                int timeTaken = q.front().second;
                q.pop();
                int dur = informTime[head] + timeTaken;
                maxTimeTaken = max(maxTimeTaken, dur);
                for(auto it: m[head]){
                    q.push({it,dur});
                }
            }
        }
        return maxTimeTaken;
    }
};

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.py
class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        q = deque([(headID,0)])
        subordinates = defaultdict(list)
        res = 0
        
        for i,v in enumerate(manager):
            subordinates[v].append(i)
            
        while q:
            u, time = q.popleft()
            res = max(res, time)
            for v in subordinates[u]:
                q.append((v, time + informTime[u]))
        return res